sql-plugin/src/main/scala/com/nvidia/spark/rapids/SplitColumnarBatchProcessor.scala

```diff
@@ -0,0 +1,333 @@
+   */
+  def addPartitionValuesToBatch(
+      batch: ColumnarBatch,
+      partitionRows: Array[Long],
+      partitionValues: Array[InternalRow],
+      partitionSchema: StructType): Iterator[ColumnarBatch] = {
+    if (partitionSchema.nonEmpty) {
+      val partitionedGroups = splitPartitionDataIntoGroups(partitionRows, partitionValues,
+        partitionSchema)
+      splitAndCombineBatchWithPartitionData(batch, partitionedGroups, partitionSchema)
+    } else {
+      new SingleGpuColumnarBatchIterator(batch)
+    }
+  }
+
+  /***
+   * Adds a single set of partition values to all rows in a ColumnarBatch.
+   * @return a new columnar batch iterator with partition values
+   */
+  def addSinglePartitionValueToBatch(
+      batch: ColumnarBatch,
+      partitionValues: InternalRow,
+      partitionSchema: StructType): Iterator[ColumnarBatch] = {
+    addPartitionValuesToBatch(batch, Array(batch.numRows), Array(partitionValues), partitionSchema)
+  }
+
+  /**
+   * Splits partitions into smaller batches, ensuring that the batch size
+   * for each column does not exceed the specified cudfLimit.
+   *
+   * Data structures:
+   * - sizeOfBatch: HashMap that stores the size of batches for each column.
+   * - currentBatch: Array used to hold the rows and partition values of the current batch.
+   * - resultBatches: Array that stores the resulting batches after splitting.
+   *
+   * Algorithm:
+   * - Initialize `sizeOfBatch` and `resultBatches`.
+   * - Iterate through `partRows`:
+   *   - Add (rows, [values]) to the current batch.
+   *   - For each column:
+   *     - Calculate the size of the current partition for the column.
+   *     - Check if adding this partition to the current batch would exceed cuDF limit.
+   *       - If it does, split the partition:
+   *         - Update the current batch's last value = (new rows, [values]).
+   *         - Add it to the result batches.
+   *         - Process the remaining rows again.
+   *       - Otherwise, increase the `sizeOfBatch` for the column.
+   *
+   * Example:
+   * Input:
+   *  partition rows:   [10, 40, 70, 10, 11]
+   *  partition values: [ [abc, ab], [bc, ab], [abc, bc], [aa, cc], [ade, fd] ]
+   *  cuDF limit:  300 bytes
+   *
+   * Result:
+   * [
+   *  [ (10, [abc, ab]), (40, [bc, ab]), (63, [abc, bc]) ],
+   *  [ (7, [abc, bc]), (10, [aa, cc]), (11, [ade, fd]) ]
+   * ]
+   *
+   * @return An array of batches, containing (row counts, partition values) pairs,
+   *         such that each batch's size is less than cudfLimit.
+   */
+  private def splitPartitionDataIntoGroups(partRows: Array[Long], partValues: Array[InternalRow],
+      partSchema: StructType): Array[Array[(Long, InternalRow)]] = {
+    val resultBatches = ArrayBuffer[Array[(Long, InternalRow)]]()
+    val currentBatch = ArrayBuffer[(Long, InternalRow)]()
+    // Initialize a HashMap to store the size of the current batch for each column
+    val sizeOfBatch = mutable.HashMap.empty[Int, Long]
```

Comment:
I'm not sure why a hash map is being used here.  The keys are all indices, so why not use an Array[Long] to store the size of each column in the current batch?

Reply:
This comment was not addressed, still curious.

Reply:
Odd, didn't see you comments posted until just now.  The overhead of using a HashMap will dwarf any "savings" by avoiding other partition columns.  Massive numbers of partition columns is not a real-world thing, as that would fracture data into super tiny files.